Adding subfeatures guidance (#17352)
mdn/browser-compat-data: docs/data-guidelines.md

-- END TEMPLATE -->

## Adding subfeatures

> Subfeatures should only be added to indicate a  _difference_ in browser compatibility versions from the parent feature; they should not be added just to indicate that a particular feature exists.
> 

Author: ```suggestion
Subfeatures should only be added to indicate a  _difference_ in browser compatibility versions from the parent feature; they should not be added just to indicate that a particular feature exists.

As a rule of thumb, if a subfeature would have the same versions as its parent feature, then it should not be added.

```

> For example, the `ImageData` interface was initially defined with `data`, `height` and `width` properties.

Author: This is not actually true ImageData has these values defined in BCD. We should remove them right @queengooborg ?

FWIW I understand the reason for not documenting "to indicate support", but it does make it hard to know whether you  have just omitted something or there is no known compatibility break.

Reviewer: It's actually the opposite: all subfeatures should be added regardless of whether they match the parent version!  Each subfeature should be added to BCD, even if the data matches its parent, to help with clarity.  Say `ImageData` has recently added a new `format` subfeature -- consumers of BCD wouldn't be able to tell whether or not `format` has been supported if the assumption is that all undefined subfeatures match the parent.

Behavioral differences, on the other hand, are more subjective.  If the spec has changed the behavior of a feature and browsers needed to update, then we should add a subfeature for that behavior change.

There are a number of subfeatures we are missing from BCD at the moment, which I'm hoping to fix in the future!

Author: @queengooborg OK. That makes sense. It concerns me that this was not communicated to me. Or the new way of adding bug URLs. 

I'll try an update then.

Author: P.S. We could reduce duplication by allowing `"support": mirror` for the whole compat section of a subfeature implemented at the same versions as its parent.

> For example, the `ImageData` interface was initially defined with `data`, `height` and `width` properties.
> As these properties were defined in the same versions as the parent feature [`ImageData`](https://github.com/mdn/browser-compat-data/blob/main/api/ImageData.json), they do not require a subfeature entry.
> However the `colorSpace` property, the constructor, and support for using the API in workers were added in later versions.
> This indicates a compatibility break, which is why a subfeature entry is required.

Author: ```suggestion
For example, the `ImageData` interface was initially defined with `data`, `height` and `width` properties.
These do not require a subfeature entry because they were defined in the same versions as the parent feature [`ImageData`](https://github.com/mdn/browser-compat-data/blob/main/api/ImageData.json).

The subfeatures for the `colorSpace` property, the constructor, and support for using the API in workers, are required, because they appear in different versions than the original `ImageData` feature.
```

-- END TEMPLATE -->

> ## Adding subfeatures

Author: @queengooborg @foolip OK, I have reworded this to be the opposite. This says "always add a subfeature for any entity or behaviour where there is an implementation in a browser."

Subfeatures should be added for all methods, properties, tags, attributes, method arguments, and other specification entities, provided that they have an implementation in one or more of the supported browser engines.

> In addition, subfeatures should be added to indicate support for specific behaviours, such as support in workers, or that and interface is serializable.

Author: ```suggestion
In addition, subfeatures should be added to indicate support for specific behaviours, such as support in workers, or that an interface is serializable.
```

-- END TEMPLATE -->

> ## Adding features and subfeatures

Author: @queengooborg Updated as per our discussion:

- features required for all released web platform features. Term feature is now defined.
- subfeatures term now defined
- subfeature required for interface methods and properties by default if there is any released implementation.
- subfeatures not requried for anything else unless there is a change in version support. This has nothing to do with what the spec is doing and is entirely due to some implementation change.

I think this is at least close, though there might be other exceptions like interface methods and properties. See comment below.

There is also the slight problem of things like HTTP headers which are not necessarily implemented in a browser, and for which the term "supported" is very loose. I think we should get this in and then address that separately.

A json file should be created for every feature that has a released public implementation on one of the supported set of browsers (and not before).
The top level entry in the file indicates the supported browser versions for the feature.

A feature may contain _subfeatures_, which document the browser versions in which parts of the parent feature were first supported or removed.
For example, a method subfeature entry would indicate the browser versions in which the method was added to the parent interface.
Subfeatures are also used to reflect changes in behaviour in particular versions, such as "support in workers" or "returns a promise".
A subfeature might nest its own subfeatures, and so on. For example, an interface feature might nest a method subfeature, which nests an option parameter, which then nests a subfeature for a value of that option.

Subfeatures entries are added for all:

- interface methods and properties that have an implementation in any of the supported browsers.

> - ?

Author: Are there an other exceptions?
